src/main/kotlin/com/team4099/robot2021/subsystems/Drivetrain.kt

```diff
@@ -0,0 +1,276 @@
+    Logger.addSource("Drivetrain", "Back Left Wheel Angles") { wheelAngles[2] }
+    Logger.addSource("Drivetrain", "Back Right Wheel Angles") { wheelAngles[3] }
+
+    // Wheel positions (from odometry)?
+    Logger.addSource("Drivetrain", "Front Left Wheel Position") { frontLeftWheelLocation }
+    Logger.addSource("Drivetrain", "Front Right Wheel Position") { frontRightWheelLocation }
+    Logger.addSource("Drivetrain", "Back Left Wheel Position") { backLeftWheelLocation }
+    Logger.addSource("Drivetrain", "Back Right Wheel Position") { backRightWheelLocation }
+
+    //  gyro angle
+    Logger.addSource("Drivetrain", "Gyro Angle") { gyroAngle }
+    //  pathfollow time stamp (need to do pathfollow stuff first?)
+
+    //  if gyro is connected boolean
+    Logger.addSource("Drivetrain", "Gyro Connected") {  }
+
+    // TODO: X Pose and Y pose, wheel positions, wheel velocity, wheel target velocities?
+
+  }
+
+  /**
+   * Sets the drivetrain to the specified angular and X & Y velocities.
+   * Calculates angular and linear velocities and calls set for each Wheel object.
+   *
+   * @param angularVelocity The angular velocity of a specified drive
+   * @param driveVector.first The linear velocity on the X axis
+   * @param driveVector.second The linear velocity on the Y axis
+   *
+   */
+  fun set(angularVelocity: AngularVelocity, driveVector: Pair<LinearVelocity, LinearVelocity>) {
+    val vX = if (isFieldOriented) {
+      driveVector.first * gyroAngle.cos -
+        driveVector.second * gyroAngle.sin
+    } else {
+      driveVector.first
+    }
+    val vY = if (isFieldOriented) {
+      driveVector.second * gyroAngle.cos +
+        driveVector.first * gyroAngle.sin
+    } else {
+      driveVector.second
+    }
+    val a =
+      vX - angularVelocity * Constants.Drivetrain.DRIVETRAIN_LENGTH / 2
+    val b =
+      vX + angularVelocity * Constants.Drivetrain.DRIVETRAIN_LENGTH / 2
+    val c =
+      vY - angularVelocity * Constants.Drivetrain.DRIVETRAIN_WIDTH / 2
+    val d =
+      vY + angularVelocity * Constants.Drivetrain.DRIVETRAIN_WIDTH / 2
+
+    wheelSpeeds[0] = hypot(b, d)
+    wheelSpeeds[1] = hypot(b, c)
+    wheelSpeeds[2] = hypot(a, d)
+    wheelSpeeds[3] = hypot(a, c)
+
+    val maxWheelSpeed = wheelSpeeds.max()
+    if (maxWheelSpeed != null && maxWheelSpeed > Constants.Drivetrain.DRIVE_SETPOINT_MAX) {
+      for (i in 0 until Constants.Drivetrain.WHEEL_COUNT) {
+        wheelSpeeds[i] = wheelSpeeds[i] / maxWheelSpeed.inMetersPerSecond
+      }
+    }
+    wheelAngles[0] = atan2(b, d)
+    wheelAngles[1] = atan2(b, c)
+    wheelAngles[2] = atan2(a, d)
+    wheelAngles[3] = atan2(a, c)
+
+    wheels[0].set(wheelAngles[0], wheelSpeeds[0])
+    wheels[1].set(wheelAngles[1], wheelSpeeds[1])
+    wheels[2].set(wheelAngles[2], wheelSpeeds[2])
+    wheels[3].set(wheelAngles[3], wheelSpeeds[3])
+
+    // odometry
+    // TODO: Put in periodic
+    var gyro2d = Rotation2d(gyroAngle.inRadians)
```

Comment:
i thought we added a thing for angle.toRotation, if not we should so we get rid of this pattern